path_login.go

```diff
@@ -0,0 +1,239 @@
+package jwtauth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	oidc "github.com/coreos/go-oidc"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/helper/cidrutil"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+func pathLogin(b *jwtAuthBackend) *framework.Path {
+	return &framework.Path{
+		Pattern: `login$`,
+		Fields: map[string]*framework.FieldSchema{
+			"role": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The role to log in against.",
+			},
+			"token": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The signed JWT to validate.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLogin,
+		},
+
+		HelpSynopsis:    pathLoginHelpSyn,
+		HelpDescription: pathLoginHelpDesc,
+	}
+}
+
+func (b *jwtAuthBackend) pathLogin(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	token := d.Get("token").(string)
+	if len(token) == 0 {
+		return logical.ErrorResponse("missing token"), nil
+	}
+
+	roleName := d.Get("role").(string)
+	if len(roleName) == 0 {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse("role could not be found"), nil
+	}
+
+	if req.Connection != nil && !cidrutil.RemoteAddrIsOk(req.Connection.RemoteAddr, role.BoundCIDRs) {
+		return logical.ErrorResponse("request originated from invalid CIDR"), nil
+	}
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	// Here is where things diverge. If it is using OIDC discovery, validate
+	// that way; otherwise validate against the locally configured keys. Once
+	// things are validated, we re-unify the request path when evaluating the
+	// claims.
+	allClaims := map[string]interface{}{}
+	switch {
+	case len(config.ParsedJWTPubKeys) != 0:
```

Comment:
Did you look at creating a static key set in the `go-oidc` library used below?  I think it might allow you to consolidate the validation library across these two methods.

Reply:
I didn't, but will

Reply:
The interface sadly doesn't allow for providing just keys but rather requires you to actually do the full verification, so it's no easier. Will keep this as-is for now.